src/cephadm/tests/test_nfs.py

```diff
@@ -0,0 +1,564 @@
+        nfsg = cephadm.NFSGanesha(
+            ctx,
+            SAMPLE_UUID,
+            "fred",
+            good_nfs_json(),
+        )
+
+        with mock.patch("cephadm.call") as _call:
+            _call.return_value = ("NFS-Ganesha Release = V100", "", 0)
+            ver = nfsg.get_version(ctx, "fake_version")
+            _call.assert_called()
+        assert ver == "100"
+
+
+def test_nfsganesha_get_daemon_name():
+    with with_cephadm_ctx([]) as ctx:
+        nfsg = cephadm.NFSGanesha(
+            ctx,
+            SAMPLE_UUID,
+            "fred",
+            good_nfs_json(),
+        )
+        assert nfsg.get_daemon_name() == "nfs.fred"
+
+
+def test_nfsganesha_get_container_name():
+    with with_cephadm_ctx([]) as ctx:
+        nfsg = cephadm.NFSGanesha(
+            ctx,
+            SAMPLE_UUID,
+            "fred",
+            good_nfs_json(),
+        )
+        name1 = nfsg.get_container_name()
+        assert name1 == "ceph-2d018a3f-8a8f-4cb9-a7cf-48bebb2cbaae-nfs.fred"
+        name2 = nfsg.get_container_name(desc="extra")
+        assert (
+            name2 == "ceph-2d018a3f-8a8f-4cb9-a7cf-48bebb2cbaae-nfs.fred-extra"
+        )
+
+
+def test_nfsganesha_get_daemon_args():
+    with with_cephadm_ctx([]) as ctx:
+        nfsg = cephadm.NFSGanesha(
+            ctx,
+            SAMPLE_UUID,
+            "fred",
+            good_nfs_json(),
+        )
+        args = nfsg.get_daemon_args()
+        assert args == ["-F", "-L", "STDERR"]
+
+
+@mock.patch("cephadm.logger")
+def test_nfsganesha_create_daemon_dirs(_logger, cephadm_fs):
+    with with_cephadm_ctx([]) as ctx:
+        nfsg = cephadm.NFSGanesha(
+            ctx,
+            SAMPLE_UUID,
+            "fred",
+            good_nfs_json(),
+        )
+        with pytest.raises(OSError):
+            nfsg.create_daemon_dirs("/var/tmp", 45, 54)
+        cephadm_fs.create_dir("/var/tmp")
+        nfsg.create_daemon_dirs("/var/tmp", 45, 54)
+        # TODO: make assertions about the dirs created
+
+
+@mock.patch("cephadm.logger")
+def test_nfsganesha_create_daemon_dirs_rgw(_logger, cephadm_fs):
+    with with_cephadm_ctx([]) as ctx:
+        nfsg = cephadm.NFSGanesha(
+            ctx,
+            SAMPLE_UUID,
+            "fred",
+            nfs_json(pool=True, files=True, rgw=True),
+        )
+        cephadm_fs.create_dir("/var/tmp")
+        nfsg.create_daemon_dirs("/var/tmp", 45, 54)
+        # TODO: make assertions about the dirs created
+
+
+@pytest.mark.parametrize(
+    "args",
+    [
+        (["foobar", "wilma", good_haproxy_json(), SAMPLE_HAPROXY_IMAGE]),
+        ([SAMPLE_UUID, "", good_haproxy_json(), SAMPLE_HAPROXY_IMAGE]),
+        ([SAMPLE_UUID, "wilma", good_haproxy_json(), ""]),
+        (
+            [
+                SAMPLE_UUID,
+                "wilma",
+                haproxy_json(files=False),
+                SAMPLE_HAPROXY_IMAGE,
+            ]
+        ),
+    ],
+)
+def test_haproxy_validation_errors(args):
```

Comment:
any particular reason for also including the haproxy and keepalive tests here? Just thinking down the line if I wanted to look at a test for haproxy or keepalived I don't think I would intuitively think to check in a `test_nfs` file.

Reply:
Done.